// @microsoft/fast-components/dist/esm/text-field/index.js
import { __decorate } from "tslib";
import { attr, customElement } from "@microsoft/fast-element";
import { TextFieldTemplate as template, TextField } from "@microsoft/fast-foundation";
import { TextFieldStyles as styles } from "./text-field.styles";
/**
 * The FAST Text Field Custom Element. Implements {@link @microsoft/fast-foundation#TextField},
 * {@link @microsoft/fast-foundation#TextFieldTemplate}
 *
 *
 * @public
 * @remarks
 * HTML Element: \<fast-text-field\>
 *
 * {@link https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot/delegatesFocus | delegatesFocus}
 */
let FASTTextField = class FASTTextField extends TextField {
    /**
     * @internal
     */
    appearanceChanged(oldValue, newValue) {
        if (oldValue !== newValue) {
            this.classList.add(newValue);
            this.classList.remove(oldValue);
        }
    }
    /**
     * @internal
     */
    connectedCallback() {
        super.connectedCallback();
        if (!this.appearance) {
            this.appearance = "outline";
        }
    }
};
__decorate([
    attr
], FASTTextField.prototype, "appearance", void 0);
FASTTextField = __decorate([
    customElement({
        name: "fast-text-field",
        template,
        styles,
        shadowOptions: {
            delegatesFocus: true,
        },
    })
], FASTTextField);
export { FASTTextField };
/**
 * Styles for TextField
 * @public
 */
export const TextFieldStyles = styles;
