// @microsoft/fast-components/dist/esm/tabs/fast-tabs.definition.js
import { Orientation } from "@microsoft/fast-web-utilities";
import { DataType } from "@microsoft/fast-tooling";
export const fastTabsDefinition = {
    version: 1,
    tags: [
        {
            name: "fast-tabs",
            description: "The FAST tabs element",
            attributes: [
                {
                    name: "orientation",
                    description: "The orientation attribute",
                    default: Orientation.horizontal,
                    required: false,
                    type: DataType.string,
                    values: [
                        { name: Orientation.horizontal },
                        { name: Orientation.vertical },
                    ],
                },
                {
                    name: "activeid",
                    description: "The activeid attribute",
                    default: undefined,
                    required: false,
                    type: DataType.string,
                },
            ],
            slots: [
                {
                    name: "tab",
                    description: "The tab slot",
                },
                {
                    name: "tabpanel",
                    description: "The tabpanel slot",
                },
                {
                    name: "start",
                    description: "The start slot",
                },
                {
                    name: "end",
                    description: "The start slot",
                },
            ],
        },
    ],
};
