// @microsoft/fast-components/dist/esm/button/index.js
import { __decorate } from "tslib";
import { attr, customElement } from "@microsoft/fast-element";
import { Button, ButtonTemplate as template } from "@microsoft/fast-foundation";
import { ButtonStyles as styles } from "./button.styles";
/**
 * The FAST Button Element. Implements {@link @microsoft/fast-foundation#Button},
 * {@link @microsoft/fast-foundation#ButtonTemplate}
 *
 *
 * @public
 * @remarks
 * HTML Element: \<fast-button\>
 *
 * {@link https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot/delegatesFocus | delegatesFocus}
 */
let FASTButton = class FASTButton extends Button {
    appearanceChanged(oldValue, newValue) {
        if (oldValue !== newValue) {
            this.classList.add(newValue);
            this.classList.remove(oldValue);
        }
    }
    connectedCallback() {
        super.connectedCallback();
        if (!this.appearance) {
            this.appearance = "neutral";
        }
    }
};
__decorate([
    attr
], FASTButton.prototype, "appearance", void 0);
FASTButton = __decorate([
    customElement({
        name: "fast-button",
        template,
        styles,
        shadowOptions: {
            delegatesFocus: true,
        },
    })
], FASTButton);
export { FASTButton };
/**
 * Styles for Button
 * @public
 */
export const ButtonStyles = styles;
