// @microsoft/fast-components/dist/esm/badge/fast-badge.definition.js
import { DataType } from "@microsoft/fast-tooling";
export const fastBadgeDefinition = {
    version: 1,
    tags: [
        {
            name: "fast-badge",
            description: "The FAST badge element",
            attributes: [
                {
                    name: "circular",
                    description: "The circular attribute",
                    type: DataType.string,
                    default: "",
                    required: false,
                },
                {
                    name: "fill",
                    description: "The fill attribute",
                    type: DataType.string,
                    default: undefined,
                    required: false,
                },
                {
                    name: "color",
                    description: "The color attribute",
                    type: DataType.string,
                    default: undefined,
                    required: false,
                },
            ],
            slots: [
                {
                    name: "",
                    description: "The default slot",
                },
            ],
        },
    ],
};
