// @microsoft/fast-components/dist/esm/anchor/index.js
import { __decorate } from "tslib";
import { attr, customElement } from "@microsoft/fast-element";
import { Anchor, AnchorTemplate as template } from "@microsoft/fast-foundation";
import { AnchorStyles as styles } from "./anchor.styles";
/**
 * The FAST Anchor Element. Implements {@link @microsoft/fast-foundation#Anchor},
 * {@link @microsoft/fast-foundation#AnchorTemplate}
 *
 *
 * @public
 * @remarks
 * HTML Element: \<fast-anchor\>
 *
 * {@link https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot/delegatesFocus | delegatesFocus}
 */
let FASTAnchor = class FASTAnchor extends Anchor {
    appearanceChanged(oldValue, newValue) {
        if (oldValue !== newValue) {
            this.classList.add(newValue);
            this.classList.remove(oldValue);
        }
    }
    connectedCallback() {
        super.connectedCallback();
        if (!this.appearance) {
            this.appearance = "neutral";
        }
    }
};
__decorate([
    attr
], FASTAnchor.prototype, "appearance", void 0);
FASTAnchor = __decorate([
    customElement({
        name: "fast-anchor",
        template,
        styles,
        shadowOptions: {
            delegatesFocus: true,
        },
    })
], FASTAnchor);
export { FASTAnchor };
/**
 * Styles for Anchor
 * @public
 */
export const AnchorStyles = styles;
